test/integration/scheduler: fix data races

pohly opened the pull request:
#### What type of PR is this?

/kind bug
/kind failing-test

#### What this PR does / why we need it:

The plugins get called by scheduler goroutines. At least the polling seems to be done concurrently and thus needs atomic access.

Locking the PreBindPlugin state is less obvious. It might be that the scheduler is really done with the test pod, but that ordering doesn't seem to be enough for the race detector. It's simpler to add mutex locking.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

pohly committed:
test/integration/scheduler: fix data races

The plugins get called by scheduler goroutines. At least the polling seems to
be done concurrently and thus needs locking.

Locking the PreBindPlugin state is less obvious. It might be that the scheduler
is really done with the test pod, but that ordering doesn't seem to be enough
for the race detector. It's simpler to add mutex locking.

pohly commented:
/retest

kerthcet commented on test/integration/scheduler/plugins/plugins_test.go:
```diff
@@ -248,12 +253,18 @@ func (sp *ScoreWithNormalizePlugin) Name() string {
 
 // reset returns name of the score plugin.
 func (sp *ScoreWithNormalizePlugin) reset() {
+	sp.mutex.Lock()
```
The method will be removed in https://github.com/kubernetes/kubernetes/pull/116634, but I think the conflict is easy to solve.

kerthcet commented on test/integration/scheduler/plugins/plugins_test.go:
```diff
@@ -903,12 +940,14 @@ func TestNormalizeScorePlugin(t *testing.T) {
 		t.Errorf("Expected the pod to be scheduled. error: %v", err)
 	}
 
+	scoreWithNormalizePlugin.mutex.Lock()
```
If pod is scheduled, I think no need of the mutex anymore, the only update to the `numScoreCalled` is scoring, and we can ignore reset here.

pohly replied:
I prefer to be thorough and do locking even if not needed. That's much easier to review for correctness than "that other code shouldn't get called in parallel". It's not performance sensitive.

pohly replied:
So if I understand you correctly, you want to have get/set methods that handle the locking?

kerthcet replied:
Yes, any concerns?

pohly replied:
Seems overkill and for that to be effective (i.e. no way to bypass the functions and use the fields directly) the fake plugins would have to be in a different package. But I can add it.

kerthcet replied:
As clean code, I think we should for not exposing too much internal details and mutexes are now across the test cases, hard to maintain I think.

dims commented:
/assign @kerthcet 

can you please review once again?

kerthcet reviewed (commented):
FYI: I tested locally, this will lead to a 3% increase from ~322s to ~331s for `k8s.io/kubernetes/test/integration/scheduler/plugins` integration tests.

pohly commented on test/integration/scheduler/plugins/plugins_test.go:
```diff
@@ -76,23 +75,62 @@ type PreFilterPlugin struct {
 }
 
+func (sp *ScorePlugin) deepCopy() *ScorePlugin {
+	sp.mutex.Lock()
```
Does it matter?

A plain mutex is simpler.

pohly commented:
> I tested locally, this will lead to a 3% increase from ~322s to ~331s for k8s.io/kubernetes/test/integration/scheduler/plugins integration tests.

Did you run multiple times to ensure that this is a statistically significant difference? I bet the variance between runs of the same code is higher than 3%, so this is probably just a coincidence.

pohly commented on test/integration/scheduler/plugins/plugins_test.go:
```diff
@@ -223,7 +335,11 @@ func (sp *ScorePlugin) reset() {
+	sp.mutex.Lock()
+	defer sp.mutex.Unlock()
+
+	curCalled := sp.numScoreCalled
```
You are right, `atomic.AddInt32` returns the new value, not the old one as I had thought.

kerthcet commented:
/retest

kerthcet commented:
> > I tested locally, this will lead to a 3% increase from ~322s to ~331s for k8s.io/kubernetes/test/integration/scheduler/plugins integration tests.
> 
> Did you run multiple times to ensure that this is a statistically significant difference? I bet the variance between runs of the same code is higher than 3%, so this is probably just a coincidence.

Actually yes, I test for several times. The results are similar.
/lgtm
Thanks @pohly

pohly commented:
@kerthcet: I squashed and rebased because of the merge conflict. Can you LGTM again?
@alculquicondor: can you approve or shall I simply lift the hold?

pohly commented:
/retest

alculquicondor commented:
From the description, it's not clear to me if you found a race in all cases or you are just trying to prevent future chances of races.

Also, the multiple deepcopies just to get one value seems excessive.

pohly commented:
> From the description, it's not clear to me if you found a race in all cases or you are just trying to prevent future chances of races.

I found multiple.

> Also, the multiple deepcopies just to get one value seems excessive.

The main advantage is simplicity in the callers.

A previous revision used `sync/atomic` for one field. Then I had to add mutex locking for some other instances where multiple values were checked. The approach with deep copies is the final iteration based on https://github.com/kubernetes/kubernetes/pull/117113#discussion_r1175198022.

pohly commented:
... final so far. I'll of course continue to revise it if requested. :sweat_smile:

alculquicondor commented on test/integration/scheduler/plugins/plugins_test.go:
```diff
@@ -216,13 +328,16 @@ func (sp *ScorePlugin) Name() string {
 // Score returns the score of scheduling a pod on a specific node.
 func (sp *ScorePlugin) Score(ctx context.Context, state *framework.CycleState, p *v1.Pod, nodeName string) (int64, *framework.Status) {
-	curCalled := atomic.AddInt32(&sp.numScoreCalled, 1)
+	sp.mutex.Lock()
```
But now all score calls are essentially serialized.

Arguably, we are calling these plugins with only a few nodes. But this is not how plugins work in production.

pohly replied:
Do you really think that mutex vs. atomic.AddInt32 makes any difference? The lock isn't going to be held long.

pohly replied:
Also remember that this is a test for correct behavior, not performance. It shouldn't depend on timing.

alculquicondor replied:
Add a comment in the plugins to clarify that they shouldn't be used for benchmarks, because they hold a lock.

pohly replied:
Instead of repeating that comment for each individual plugin, I added a package comment:
https://github.com/kubernetes/kubernetes/compare/9f556132010572c93f757a0f71b086d3743738cb..9e9a6cde4b8b446cf68704ceb6e52df7eb1e00e3

alculquicondor commented:
Can you offer more insight into where you found the races?

kerthcet commented:
> Can you offer more insight into where you found the races?

The potential race is when pod is scheduling, we'll modify the plugin fields, while the same time, we may query these fields in the tests. But honestly up to now, I didn't encounter data race errors here.

pohly commented:
> Can you offer more insight into where you found the races?

I enabled `go test -race` for all integration tests - see https://github.com/kubernetes/kubernetes/pull/116980

alculquicondor commented:
I'm not convinced that's visible enough, but maybe it's fine.

alculquicondor commented:
/lgtm

alculquicondor commented:
Feel free to release the hold if you are happy with it.

kerthcet commented:
Still LGTM, so 
/hold cancel
